File: CLDemo-OC/CLDemo-OC/Class/CLLineChartController/CLLineChartController.m
//
//  CLLineChartController.m
//  CLDemo
//
//  Created by AUG on 2019/9/3.
//  Copyright © 2019 JmoVxia. All rights reserved.
//

#import "CLLineChartController.h"
#import "CLLineChartView.h"
#import <Masonry/Masonry.h>

@interface CLLineChartController ()<CLLineChartViewDataSource>

///折线图
@property (nonatomic, strong) CLLineChartView *chart;
///数据源
@property (nonatomic, strong) NSMutableArray<CLLineChartPoint *> *arrayDS;

@end

@implementation CLLineChartController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.chart = [[CLLineChartView alloc] init];
    self.chart.dataSource = self;
    [self.view addSubview:self.chart];
    [self.chart mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.mas_equalTo(0);
        make.width.mas_equalTo(250);
        make.height.mas_equalTo(200);
    }];
    
    self.arrayDS = [NSMutableArray array];
    
    for (NSInteger i = 0; i < 24; i++) {
        CLLineChartPoint *point = [CLLineChartPoint new];
        point.x = i + 0.5;
        point.y = [self randomBetween:7000 AndBigNum:8000 AndPrecision:100];
        [self.arrayDS addObject:point];
    }
    [self.chart reload];
}
/**
 *   samllNum:  两数中的最小值
 *   bigNum: 两数中的最大值
 *   precision: 精度值，如：精确1位小数，precision参数值为10； 两位小数precision参数值为100；
 */
- (float)randomBetween:(float)smallNum AndBigNum:(float)bigNum AndPrecision:(NSInteger)precision{
    //求两数之间的差值
    float subtraction = bigNum - smallNum;
    //取绝对值
    subtraction = ABS(subtraction);
    //乘以精度的位数
    subtraction *= precision;
    //在差值间随机
    float randomNumber = arc4random() % ((int) subtraction + 1);
    //随机的结果除以精度的位数
    randomNumber /= precision;
    //将随机的值加到较小的值上
    float result = MIN(smallNum, bigNum) + randomNumber;
    //返回结果
    return result;
}
- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [super touchesBegan:touches withEvent:event];
    [self.arrayDS removeAllObjects];
    for (NSInteger i = 0; i < 24; i++) {
        CLLineChartPoint *point = [CLLineChartPoint new];
        point.x = i + 0.5;
        point.y = [self randomBetween:7000 AndBigNum:8000 AndPrecision:100];
        [self.arrayDS addObject:point];
    }
    [self.chart updateWithConfigure:^(CLLineChartConfigure * _Nonnull configure) {
        UIColor *color = [UIColor colorWithHue:(arc4random() % (256) / (256.0)) saturation:(arc4random() % (256) / (256.0)) brightness:(arc4random() % (256) / (256.0)) alpha:(1.0)];
        configure.lineWidth = 0.5;
        configure.lineColor = color;
        configure.gradientColors = @[(__bridge id)[color colorWithAlphaComponent:0.5].CGColor,(__bridge id)[color colorWithAlphaComponent:0.0].CGColor];
        configure.dottedLineColor = color;
        configure.dottedLineWidth = 0.5;
    }];
    [self.chart reload];
}
///点数据
- (NSArray<CLLineChartPoint *> *)lineChartViewPoints {
    return self.arrayDS;
}
///x最小
- (CGFloat)lineChartViewXLineMin {
    return 0;
}
///x最大
- (CGFloat)lineChartViewXLineMax {
    return 24;
}
///y最小
- (CGFloat)lineChartViewYLineMin {
    return 7000;
}
///y最大
- (CGFloat)lineChartViewYLineMax {
    return 8000;
}
///虚线位置
- (CGFloat)lineChartViewDottedLineY {
    return [self randomBetween:7000 AndBigNum:8000 AndPrecision:100];
}
///图表宽高
- (CGSize)lineChartViewChartSize {
    return CGSizeMake(250, 200);
}
- (void)dealloc
{
    
}
@end

File: CLDemo-Swift/CLDemo-Swift/Class/CLLineChartController/CLLineChartController.swift
//
//  CLLineChartController.swift
//  CLDemo-Swift
//
//  Created by JmoVxia on 2022/11/11.
//

import UIKit

class CLLineChartController: CLController {
    lazy var lineChartView: CLLineChartView = {
        let view = CLLineChartView()
        view.dataSource = self
        view.backgroundColor = .random
        return view
    }()

    var points = [CGPoint]()
    var points1 = [CGPoint]()

    override func viewDidLoad() {
        super.viewDidLoad()
        view.addSubview(lineChartView)
        lineChartView.snp.makeConstraints { make in
            make.center.equalToSuperview()
            make.size.equalTo(CGSize(width: 250, height: 150))
        }

        for i in 0 ..< 40 {
            points.append(.init(x: CGFloat(i) + 0.5, y: CGFloat.random(in: 7000 ... 7500)))
            points1.append(.init(x: CGFloat(i) + 0.5, y: CGFloat.random(in: 7500 ... 8000)))
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            self.lineChartView.reload()
        }
    }
}

extension CLLineChartController: CLLineChartViewDataSource {
    func chartView(_ chartView: CLLineChartView, gradientColorsForLineAt line: Int) -> [CGColor] {
        ["#00FFFF".cgColor, "#00FFFF".uiColor.withAlphaComponent(0.0).cgColor]
    }

    func numberOfXTicks(in view: CLLineChartView) -> Int {
        6
    }

    func numberOfYTicks(in view: CLLineChartView) -> Int {
        6
    }

    func numberOfLines(in view: CLLineChartView) -> Int {
        1
    }

    func chartView(_ chartView: CLLineChartView, numberOfPointsInLine line: Int) -> [CGPoint] {
        if line == 0 {
            points
        } else {
            points1
        }
    }

    func chartView(_ chartView: CLLineChartView, layerForLineAt line: Int) -> CAShapeLayer {
        let layer = CAShapeLayer()
        layer.lineWidth = 1
        layer.fillColor = UIColor.clear.cgColor
        layer.strokeColor = "#00FFFF".cgColor
        return layer
    }

    func axisInX(in view: CLLineChartView) -> (min: CGFloat, max: CGFloat) {
        (0, 40)
    }

    func axisInY(in view: CLLineChartView) -> (min: CGFloat, max: CGFloat) {
        (7000, 8000)
    }

    func size(in view: CLLineChartView) -> CGSize {
        CGSize(width: 250, height: 150)
    }
}

extension CLLineChartController {
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        points.removeAll()
        points1.removeAll()
        for i in 0 ..< 40 {
            points.append(.init(x: CGFloat(i) + 0.5, y: CGFloat.random(in: 7000 ... 7500)))
            points1.append(.init(x: CGFloat(i) + 0.5, y: CGFloat.random(in: 7500 ... 8000)))
        }
        lineChartView.reload()
    }
}
